src/generate_features.py

```diff
@@ -1,3 +1,215 @@
+    logger.info('Clustering GMM with n_components=%d and cov_type=%s',
+                n_components, cov_type)
+    gmm = GMM(n_components=n_components, covariance_type=cov_type)
+    gmm.fit(X)
+    labels = gmm.predict(X)
+    return labels
+
+
+def km_clust(X, k):
+    logger.debug('Clustering KMeans with k=%d', k)
+    km = KMeans(X, k)
+    km.fit(X)
+    labels = km.labels_
+    return labels
+
+
+def gen_plots(order_types):
+    # TODO: implement heatmap plotting and save imgs
+    pass
+
+
+if __name__ == '__main__':
+    order_products = load_data(data_dir, 'order_products__train.csv')
+    order_products_prior = load_data(data_dir, 'order_products__prior.csv')
+    orders = load_data(data_dir, 'orders.csv')
+    products = load_data(data_dir, 'products.csv')
+    aisles = load_data(data_dir, 'aisles.csv')
+    departments = load_data(data_dir, 'departments.csv')
+
+    products.set_index('product_id', inplace=True)
+    orders.set_index('order_id', inplace=True)
+    aisles.set_index('aisle_id', inplace=True)
+    departments.set_index('department_id', inplace=True)
+
+    order_products = pd.concat([order_products, order_products_prior])
+    del order_products_prior
+
+    n_orders = order_products.groupby('product_id').size().rename('n_orders')
+    products = products.join(n_orders)
+
+    logger.debug('Engineering product features')
+    # Product considered "popular" if among the top_n products ordered in its department
+    top_n = config.get('popular_threshold', 5)
+
+    top_products = products.groupby('department_id')['n_orders']\
+                   .nlargest(top_n).reset_index()
+    top_products = top_products.join(products[['product_name']], on='product_id')
+    top_products = top_products.values.reshape(-1)
+    products['popular'] = products['product_name'].isin(top_products).astype(int)
+
+    products['organic'] = products['product_name'].str.lower()\
+                          .str.contains('organic').astype(int)
+
+    # cats contains mapping of macro-level category (vegetable, beverage, etc) to
+    # all aisles which could fall under that classification
+    for cat in cats:
+        aisles[cat] = aisles['aisle'].isin(cat_map[cat]).astype(int)
+
+    logger.debug('Engineering order features')
+    orders = orders.join(order_products['order_id'].value_counts().rename('order_size'))
+
+    # discretize order size to help with long-tail issue
+    # these categories not currently used
+    orders['size_cat'] = pd.cut(orders['order_size'],
+                                [0, 5, 10, 20, np.inf],
+                                labels=['small', 'medium', 'large', 'xl'])
+
+    full = order_products.join(orders, on='order_id')
+    full = full.join(products, on='product_id')
+    full = full.join(aisles, on='aisle_id')
+
+    groups = full.groupby('order_id')
+
+    # order_types will be combination of metadata and product-level features
+    order_types = groups[cats + ['reordered', 'organic', 'popular']].mean()
```

Comment:
why are some lists in config and some hard coded here

Reply:
`cats` file maps aisles to product categories and are static by nature, `reordered`, `organic`, and `popular` are based on the data itself and is dynamic (popular depends on the universe of orders, and reordered depends on a user's activity. organic is technically just based on product's name but this is simpler to code than enumerate)